### 1. Парсинг новостей

Для парсинга новостей был выбран портал Lenta.ru, особенностью которого является наличие динамической кнопки "Лента добра", отфильтровывающей негативные новости. 

Для работы с динамическим элементом использовалась библиотека `Selenium`. В вызове функции предусмотрена возможность как включать, так и выключать кнопку через параметр `good_news=True`. Основной сбор данных со страницы выполняется с помощью библиотеки `BeautifulSoup`. Количество статей, которое необходимо собрать, задаётся через параметр `num_articles` функции `parse_page`.

In [1]:
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from selenium import webdriver
from selenium.webdriver.common.by import By

from bs4 import BeautifulSoup
import pandas as pd
import requests

def get_info(good_news=True, html='https://lenta.ru/parts/news/'):
    """ 
    This function retrieves information from the Lenta.ru newsfeed, with a custom option to 'show only good news'.
    """
    if good_news==True:  
        s = Service(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=s)
        driver.get(html)
        driver.implicitly_wait(3)
        driver.find_element(by=By.CLASS_NAME, value="goodnews").click()
        soup = BeautifulSoup(driver.page_source, 'html.parser') 
        
    else:
        html = requests.get('https://lenta.ru/parts/news/', verify=False)
        soup = BeautifulSoup(html.text, 'html.parser') 
    
    urls = soup.find_all('a', class_="card-full-news _parts-news")
    dict_url = {i: urls[i].get("href") for i in range(len(urls))}
    headers = soup.find_all('h3', class_="card-full-news__title")
    dict_header = {i: headers[i].get_text() for i in range(len(headers))}

    final_df = pd.DataFrame(columns=['url', 'header', 'article'])
    final_df['url'] = pd.DataFrame.from_dict(dict_url, orient='index', columns=['url'])
    final_df['header'] = pd.DataFrame.from_dict(dict_header, orient='index', columns=['header'])


    dict_article = {}
    for i in range(len(dict_url)):
        url = 'https://lenta.ru' + dict_url[i]
        response = requests.get(url)
        soup2 = BeautifulSoup(response.text, 'html.parser') 
        text = soup2.find_all('p', class_="topic-body__content-text")
        article_list = []
        for t in text:
            article = article_list.append(t.get_text())
            article = ' '.join(article_list)
        dict_article[i] = article

    final_df['article'] = pd.DataFrame.from_dict(dict_article, orient='index', columns=['article'])
    return final_df

def parse_page(good_news=True, num_articles=25):
    """ 
    The function iterates over the pages of the newsfeed to retrieve the number of articles specified in the function call.
    """
    df_final = pd.DataFrame(columns=['url', 'header', 'article'])
    
    for i in range(1,1000):
        if df_final.shape[0] < num_articles:
            n = num_articles-df_final.shape[0]
            temp_df = get_info(good_news, html='https://lenta.ru/parts/news/' + str(i))
            df_final = pd.concat([df_final, temp_df.iloc[:n]], ignore_index=True)

        else:
            return df_final

In [2]:
data = parse_page(good_news=True, num_articles=50)
data

url  \
0                          /news/2025/07/11/kultovyy/   
1   /news/2025/07/10/rossiyanka-uvidela-medvezhat-...   
2   /news/2025/07/10/rossiyskiy-region-vvedet-novy...   
3   /news/2025/07/10/dagestan-postroit-dorogu-k-sa...   
4   /news/2025/07/10/pochti-million-turistov-poset...   
5   /news/2025/07/10/v-rossii-poyavilsya-putevodit...   
6   /news/2025/07/10/nazvan-samyy-poseschaemyy-gor...   
7   /news/2025/07/10/pervaya-sumka-herm-s-birkin-u...   
8   /news/2025/07/10/rossiyskoe-vino-priznali-luch...   
9   /news/2025/07/10/v-vologde-nashli-zahoronenie-...   
10  /news/2025/07/10/devushki-rinulis-v-apteki-za-...   
11  /news/2025/07/10/lavrov-udivil-rubashkoy-na-uz...   
12    /news/2025/07/10/kim-kardashyan-smenila-imidzh/   
13  /news/2025/07/10/strimer-buster-kupil-ferrari-...   
14  /news/2025/07/10/britni-spirs-stantsevala-na-k...   
15  /news/2025/07/10/pozhilym-lyudyam-rasskazali-o...   
16  /news/2025/07/10/muzhchina-ostanovilsya-v-goro...   
17  /news/2025/07/10/nazvany-vrednye-dlya-kishechn...   
18  /news/2025/07/10/42-letnyaya-pevitsa-yuliya-ko...   
19  /news/2025/07/10/delfin-sygral-v-meduzobol-per...   
20  /news/2025/07/10/alesya-kafelnikova-pohvastala...   
21  /news/2025/07/10/pereehavshiy-v-rossiyu-portug...   
22  /news/2025/07/10/v-seti-poyavilis-snimki-46-le...   
23  /news/2025/07/10/52-letnyaya-sofiya-vergara-po...   
24  /news/2025/07/09/alena-vodonaeva-v-otkrovennom...   
25  /news/2025/07/09/novaya-zhena-milliardera-dzhe...   
26  /news/2025/07/09/stilist-lisovets-pokazal-otkr...   
27  /news/2025/07/09/devushka-pokazala-kuplennuyu-...   
28  /news/2025/07/09/kolie-v-vide-byuschegosya-ser...   
29  /news/2025/07/09/v-rossiyskom-zooparke-rodilsy...   
30  /news/2025/07/09/raskryty-pyat-samyh-modnyh-tr...   
31  /news/2025/07/09/rossiyskaya-tennisistka-poyav...   
32  /news/2025/07/09/u-beregov-rossiyskogo-regiona...   
33  /news/2025/07/09/rossiyanam-nazvali-sposob-bez...   
34  /news/2025/07/09/pyat-fasonov-obuvi-iz-80-h-ve...   
35  /news/2025/07/09/ozhestochennuyu-shvatku-krysy...   
36  /news/2025/07/09/rossiyskiy-brend-provel-pokaz...   
37  /news/2025/07/09/77-letniy-mihail-shufutinskiy...   
38  /news/2025/07/09/raskryty-sposoby-izbavitsya-o...   
39  /news/2025/07/09/moskvicham-ustroyat-yogu-s-ko...   
40  /news/2025/07/09/aktrisa-iz-kitaya-voshitilas-...   
41  /news/2025/07/09/guzeeva-pokazala-vneshnost-po...   
42  /news/2025/07/09/drevnee-zahoronenie-nashli-v-...   
43  /news/2025/07/09/keyt-middlton-vyshla-v-svet-p...   
44  /news/2025/07/09/ezh-nashel-neobychnyy-sposob-...   
45  /news/2025/07/09/priehavshiy-v-otpusk-v-rossiy...   
46  /news/2025/07/09/zhenschin-prizvali-pered-bere...   
47  /news/2025/07/09/modnyy-brend-pokazal-platie-i...   
48  /news/2025/07/09/raskryt-sekret-dostizheniya-p...   
49  /news/2025/07/09/kompaniya-mattel-vypustila-pe...   

                                               header  \
0   «Декстер» вернулся с новым сезоном и воскресши...   
1   Россиянка увидела медвежат, захотела угостить ...   
2   Российский регион введет новые правила для тур...   
3   Дагестан построит дорогу к самому большому в Е...   
4   Почти миллион туристов посетили Кабардино-Балк...   
5   В России появился путеводитель по промышленном...   
6   Назван самый посещаемый город Золотого кольца ...   
7   Первая сумка Hermès Birkin ушла на торгах за 8...   
8   Российское вино признали лучшим на конкурсе в ...   
9    В Вологде нашли захоронение второго главы города   
10          Девушки ринулись в аптеки за модной вещью   
11    Лавров удивил рубашкой на ужине в Куала-Лумпуре   
12                        Ким Кардашьян сменила имидж   
13  Стример Бустер купил Ferrari за десятки миллио...   
14  Бритни Спирс станцевала на камеру в крошечном ...   
15  Пожилым людям дали советы по профилактике глухоты   
16  Мужчина остановился в городе на одну ночь и вы...   
17     Названы вредные для кишечника ошибки в питании   
18  42-летняя певица Юлия Ковальчук поделилась сни...   
19

### 2. Суммаризация и расчет метрик

Для суммаризации текста была выбрана модель `sarahai/ruT5-base-summarizer`, основанная на архитектуре `ai-forever/ruT5-base`(разработка команды SberDevices). Для обучения модели авторами был использован датасет с 60 тысячами строк. В работе использовалась уже обученная нейронная сеть с фиксированными параметрами.

Генерация суммирующего абзаца строилась со следующими ограничениями: минимальный размер - `80` токенов, максимальный размер - `250` токенов. Устанавливая параметр `length_penalty=1.0`, мы поощряем модель давать более длинные (и, как правило, более содержательные в контексте задачи) ответы. Параметр `num_beams=4` означает, что на каждом шаге генерации рассматривается 4 наиболее вероятных вариантов продолжения последовательности. Флажок `early_stopping=False` означает, что генерация продолжается, пока есть хотя бы один потенциальный кандидат.

Для оценки суммаризации были выбраны две метрики из библиотеки `evaluate` - `rouge` и `bertscore`.

In [9]:
from transformers import T5Tokenizer, T5ForConditionalGeneration
from evaluate import load

model_name = "sarahai/ruT5-base-summarizer"  
tokenizer = T5Tokenizer.from_pretrained(model_name)
model = T5ForConditionalGeneration.from_pretrained(model_name)

rouge = load("rouge")
bertscore = load("bertscore")

for i in range(len(data)):
    input_ids = tokenizer(data['article'].values[i], return_tensors="pt").input_ids
    outputs = model.generate(input_ids, max_length=250, min_length=80, length_penalty=1.0, num_beams=4, early_stopping=False)
    data.loc[i, 'summary'] = tokenizer.decode(outputs[0], skip_special_tokens=True)

    rouge_metrics = rouge.compute(predictions=[data.loc[i, 'summary']], references=[data.loc[i, 'article']])

    data.loc[i, 'rouge1'] = rouge_metrics['rouge1']
    
    bertscore_metrics = bertscore.compute(predictions=[data.loc[i, 'summary']], references=[data.loc[i, 'article']], lang="ru")
    
    data.loc[i, 'precision'] = bertscore_metrics['precision']
    data.loc[i, 'recall'] = bertscore_metrics['recall']
    data.loc[i, 'f1'] = bertscore_metrics['f1']

### 3. Результаты

In [10]:
data

url  \
0                          /news/2025/07/11/kultovyy/   
1   /news/2025/07/10/rossiyanka-uvidela-medvezhat-...   
2   /news/2025/07/10/rossiyskiy-region-vvedet-novy...   
3   /news/2025/07/10/dagestan-postroit-dorogu-k-sa...   
4   /news/2025/07/10/pochti-million-turistov-poset...   
5   /news/2025/07/10/v-rossii-poyavilsya-putevodit...   
6   /news/2025/07/10/nazvan-samyy-poseschaemyy-gor...   
7   /news/2025/07/10/pervaya-sumka-herm-s-birkin-u...   
8   /news/2025/07/10/rossiyskoe-vino-priznali-luch...   
9   /news/2025/07/10/v-vologde-nashli-zahoronenie-...   
10  /news/2025/07/10/devushki-rinulis-v-apteki-za-...   
11  /news/2025/07/10/lavrov-udivil-rubashkoy-na-uz...   
12    /news/2025/07/10/kim-kardashyan-smenila-imidzh/   
13  /news/2025/07/10/strimer-buster-kupil-ferrari-...   
14  /news/2025/07/10/britni-spirs-stantsevala-na-k...   
15  /news/2025/07/10/pozhilym-lyudyam-rasskazali-o...   
16  /news/2025/07/10/muzhchina-ostanovilsya-v-goro...   
17  /news/2025/07/10/nazvany-vrednye-dlya-kishechn...   
18  /news/2025/07/10/42-letnyaya-pevitsa-yuliya-ko...   
19  /news/2025/07/10/delfin-sygral-v-meduzobol-per...   
20  /news/2025/07/10/alesya-kafelnikova-pohvastala...   
21  /news/2025/07/10/pereehavshiy-v-rossiyu-portug...   
22  /news/2025/07/10/v-seti-poyavilis-snimki-46-le...   
23  /news/2025/07/10/52-letnyaya-sofiya-vergara-po...   
24  /news/2025/07/09/alena-vodonaeva-v-otkrovennom...   
25  /news/2025/07/09/novaya-zhena-milliardera-dzhe...   
26  /news/2025/07/09/stilist-lisovets-pokazal-otkr...   
27  /news/2025/07/09/devushka-pokazala-kuplennuyu-...   
28  /news/2025/07/09/kolie-v-vide-byuschegosya-ser...   
29  /news/2025/07/09/v-rossiyskom-zooparke-rodilsy...   
30  /news/2025/07/09/raskryty-pyat-samyh-modnyh-tr...   
31  /news/2025/07/09/rossiyskaya-tennisistka-poyav...   
32  /news/2025/07/09/u-beregov-rossiyskogo-regiona...   
33  /news/2025/07/09/rossiyanam-nazvali-sposob-bez...   
34  /news/2025/07/09/pyat-fasonov-obuvi-iz-80-h-ve...   
35  /news/2025/07/09/ozhestochennuyu-shvatku-krysy...   
36  /news/2025/07/09/rossiyskiy-brend-provel-pokaz...   
37  /news/2025/07/09/77-letniy-mihail-shufutinskiy...   
38  /news/2025/07/09/raskryty-sposoby-izbavitsya-o...   
39  /news/2025/07/09/moskvicham-ustroyat-yogu-s-ko...   
40  /news/2025/07/09/aktrisa-iz-kitaya-voshitilas-...   
41  /news/2025/07/09/guzeeva-pokazala-vneshnost-po...   
42  /news/2025/07/09/drevnee-zahoronenie-nashli-v-...   
43  /news/2025/07/09/keyt-middlton-vyshla-v-svet-p...   
44  /news/2025/07/09/ezh-nashel-neobychnyy-sposob-...   
45  /news/2025/07/09/priehavshiy-v-otpusk-v-rossiy...   
46  /news/2025/07/09/zhenschin-prizvali-pered-bere...   
47  /news/2025/07/09/modnyy-brend-pokazal-platie-i...   
48  /news/2025/07/09/raskryt-sekret-dostizheniya-p...   
49  /news/2025/07/09/kompaniya-mattel-vypustila-pe...   

                                               header  \
0   «Декстер» вернулся с новым сезоном и воскресши...   
1   Россиянка увидела медвежат, захотела угостить ...   
2   Российский регион введет новые правила для тур...   
3   Дагестан построит дорогу к самому большому в Е...   
4   Почти миллион туристов посетили Кабардино-Балк...   
5   В России появился путеводитель по промышленном...   
6   Назван самый посещаемый город Золотого кольца ...   
7   Первая сумка Hermès Birkin ушла на торгах за 8...   
8   Российское вино признали лучшим на конкурсе в ...   
9    В Вологде нашли захоронение второго главы города   
10          Девушки ринулись в аптеки за модной вещью   
11    Лавров удивил рубашкой на ужине в Куала-Лумпуре   
12                        Ким Кардашьян сменила имидж   
13  Стример Бустер купил Ferrari за десятки миллио...   
14  Бритни Спирс станцевала на камеру в крошечном ...   
15  Пожилым людям дали советы по профилактике глухоты   
16  Мужчина остановился в городе на одну ночь и вы...   
17     Названы вредные для кишечника ошибки в питании   
18  42-летняя певица Юлия Ковальчук поделилась сни...   
19

In [11]:
data[['rouge1', 'precision', 'recall', 'f1']].mean()

rouge1       0.402223
precision    0.878469
recall       0.720545
f1           0.791524
dtype: float64

Среди доступных rouge-метрик нами была выбрана `rouge1`, которая измеряет перекрытие токенов (униграмм) в нашем оригинальном тексте и сгенерированном резюме. Недостаток этой метрики состоит в том, что изначально она предназначена для сравнения суммаризаций - произведенной человеком (эталонной) и сгенерированной моделью. В нашем варианте идеальный вариант суммаризации отсутствует, что искажает интерпретацию результатов. Так как это числовая метрика, которая просто считает отношение количества токенов в одном и другом тексте, то для большинства из статей ее значения очень малы. Например, первая статья состоит из более чем 1 тысячи токенов, тогда как для суммаризации мы установили ограничение в 250 токенов - естественным образом, мы получаем малое значение `rouge1 = 0.2963`. 

В то же время высокие показатели этой метрики мы можем наблюдать на коротких статьях, что означает, что сгенерированная суммаризация действительно состоит из слов оригинального текста. Однако нужно отметить, что метрика не принимает во внимание смысловое соответствие, а лишь дает нам возможность оценить количество совпадений слов.

Вторая метрика - `bertscore` - также оценивает сходство токенов в оригинальном и сгенерированном текстах, но принимает во внимание семантическое сходство, рассчитывая косинусное подобие между токенами.
Bertscore дает нам возможность оценить три показателя:

- `precision` - насколько сгенерированный текст соответствует содержанию оригинальной статьи;

- `recall` - насколько оригинальная статья отражена в сгенерированном тексте. Эта метрика способна оценить, были ли **упущены важные детали** из оригинальной статьи;

- `f1 score` - взвешенная метрика между precision и recall, позволяющая дать суммаризации обобщенную оценку.

Полученные результаты демонстрируют достаточно высокий уровень `precision` (среднее - 0.8785), что свидетельствует о том, что сгенерированные суммаризации в целом соответствуют содержанию оригинальных текстов. При этом значения `recall` оказываются заметно ниже (среднее - 0.7205), что может указывать на потерю части важной информации из исходных статей. Повышение обоих показателей может рассматриваться как одна из задач для дальнейшего улучшения модели суммаризации.

В процессе вычитки были обнаружены следующие **ошибки модели суммаризации**:

1) Дублирование предложений: 

*Россиянка, увидев медвежат, захотела угостить их колбаской и сняла все это на видео. Женщина обратилась к водителю автомобиля с просьбой поделиться лакомством с дикими животными. В итоге один из «медвежаток» встает на задние лапы, а другой — встает на задние лапы. В то же время один из «медвежаток» встает на задние лапы.*

*Пять фасонов популярной в 1980-х годах обуви вернулись в моду летом 2025 года. Так, актуальными вновь стали двухцветные лодочки, лодочки и лодочки. Кроме того, число трендов пополнили балетки, балетки, туфли с открытой пяткой и оксфорды. Кроме того, число трендов пополнили балетки, туфли с открытой пяткой и модели с Т-образным ремешком.*

2) Галлюцинации:

- Семантические: 

*Официальный представитель МИД России Мария Захарова раскритиковала рубашку министра иностранных дел России Сергея Лаврова на гала-ужине в Куала-Лумпуре.* (Оригинальный текст не содержит оценочного суждения внешнего вида политика).

*Актриса китайского происхождения Фэйя Цзоу, выросшая в России, поделилась мнением о Владивостоке и его жителях. По словам Цзоу, она не была никогда во Владивостоке, но очень хотела бы приехать в город, чтобы познакомиться с ним. По словам Цзоу, она занимает первое место в России по изучению китайского языка, обгоняя даже Москву и Санкт-Петербург.* (Первое место в России по изучению китайского языка занимает не девушка, а город).

- Фактические: 

*В России презентовали путеводитель по промышленному туризму. В нем туристы смогут увидеть искры доменной печи.* (В путеводителе нельзя увидеть искры, но можно узнать, где их увидеть).

*Редакторы испанской версии журнала G раскрыли пять самых модных трендов 2026 года*. (Оригинальное название журнала - GQ).


### 4. Обсуждение и дальнейшие пути изучения

В ходе выполнения задания рассматривались и другие варианты **оценки качества суммаризации**. В частности, одним из возможных подходов может стать оценка резюме по установленным параметрам (например, точность, полнота, галлюцинации) **с использованием генеративной нейросети**. Однако при тестировании этого решения были выявлены следующие проблемы:

- не все доступные генеративные модели корректно работают с русскоязычными текстами;
- большинство бесплатных моделей имеет ограничение на количество обрабатываемых токенов;
- использование модели для оценки результатов другой модели может приводить к накоплению ошибок и снижению достоверности итоговой оценки.

Кроме того, в нашем кейсе для решения задачи суммаризации была использована уже обученная модель с заранее настроенными весами. Однако в перспективе возможно не только более детальное сравнение с другими доступными моделями, но и **проведение собственного дообучения или обучения с нуля** на специально подобранном корпусе текстов. Это позволит адаптировать модель к специфике задач и языкового материала, а также потенциально повысить качество суммаризации.

### Ссылки

Gaby A. (2025). T5-Base, T5-Large, and BART — The Battle of the Summarization Models. Medium. Ссылка: https://medium.com/@gabya06/t5-vs-bart-the-battle-of-the-summarization-models-c1e6d37e56ca

Prabhat Zade (2025). ROUGE Score: A Complete Tutorial for Evaluating Text Summarization Models. Medium. Ссылка: https://medium.com/@prabhatzade/rouge-score-a-complete-tutorial-for-evaluating-text-summarization-models-a3a146417118

ruT5-base. Hugging Face. Ссылка: https://huggingface.co/ai-forever/ruT5-base

ruT5-base-summarizer. Hugging Face. Ссылка: https://huggingface.co/sarahai/ruT5-base-summarizer

Tianyi Zhang, Varsha Kishore, Felix Wu, Kilian Q. Weinberger, and Yoav Artzi. (2020). BERTScore: Evaluating Text Generation with BERT. Конференция ICLR, 2020. Ссылка: https://arxiv.org/abs/1904.09675

Transformers — Text Generation. Hugging Face Documentation. Ссылка: https://huggingface.co/docs/transformers/main_classes/text_generation